### PR TITLE
Add property wrapper-based model definitions

### PR DESCRIPTION
**Pull request by tgoyne**
This PR currently has the implementation of this but no tests. I've been testing it by modifying the existing test suite to use the new declarations and disabling all the tests which aren't applicable, but that's not a great solution so I've started on a new test suite. I figure that we can talk about the name and review the implementation while that's in progress.

Currently the property wrapper is named `@Managed`. I don't like this name very much because it's only somewhat aligned with how we're currently using the word "managed". We have managed and unmanaged objects which refers to objects in a Realm and standalone objects which aren't in a Realm, and while I'm not quite sure what "unmanaged properties" would refer to it's clearly not an analogue of unmanaged objects.

Some other ideas I had that I don't like:

 * `@ManagedProperty`: all the downsides of Managed but longer and worse.
 * `@Property`: already taken for something different.
 * `@realm`: too clever? Results in really nice looking class definitions but having both `Realm` and `realm` as types in the API is very confusing.
 * `@RealmProperty`: awfully long.
 * `@Persisted`: we very thoroughly ripped out the "persisted" term from everything and it'd be weird to use it just here.
 * `@Field`: not really a term used in Swift.
 * `@R` or `@P`: Too short?
 * `@ℜ` or `@ℜ𝔢𝖆𝖑𝖒`: I hope the problem is obvious

---
Things left to do:

 - [x] Bikeshed the name
 - [x] Add a new test suite for this
 - [x] Do final perf testing to verify that it's still faster and check if @inlineable is actually doing anything
 - [x] Changelog entry etc.
 - [x] ~~Do we actually need to be creating managed accessor subclasses? Could save a little bit of memory and startup time if not.~~ Not really, but there's some incidental dependencies on the subclasses so they can't just be skipped even though they're doing very little.

### CODE REVIEW

**Comment by nirinchev** [comment]
Not sure if my opinion weighs a lot considering I'm not a Swift developer, but I do find `Persisted` to be a more appropriate name than `Managed`. On one hand, managing a property is not really a thing, and then what does having a "managed" property on an "unmanaged" object mean.

`Persisted` is reasonably straightforward to understand and is a good juxtaposition to the "ignored" terminology we are currently using. Seeing a swift model where half the properties are annotated with `@Persisted` makes it immediately clear that those will be stored in the database, while the rest will be ignored.

`@RealmProperty` could also work despite the length - it does seem in line with the length of some of the SwiftUI property wrappers. `@RealmValue` is tangentially shorter but might be confusing in combination to `AnyRealmValue` which I believe is what the Swift version of Core's Mixed is called.

**Comment by tgoyne** [comment]
Persisted does feel like the option I have the weakest argument against.

It's a pretty silly criterion, but length-wise it'd be nice to have something shorter than `@objc dynamic` so that updating existing models to this feels like it's getting less verbose rather than the hot new thing being more clunky than what it's replacing.

**Comment by bdkjones** [comment]
Regarding the property wrapper name:

1. Of the existing options, `@realm` is the best. 

2. As a longtime Apple dev, "managed"-anything simply screams "Core Data". A large portion of folks new to Realm are probably coming from a decade+ of Core Data, so this word is loaded. I'd avoid it.

File: RealmSwift/Tests/ModernObjectTests.swift [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+        XCTAssertNotNil(list.realm)
+        XCTAssertNotNil(set.realm)
+        XCTAssertNotNil(dictionary.realm)
+        XCTAssertTrue(list === obj.arrayCol)
+        XCTAssertTrue(set === obj.setCol)
+        XCTAssertTrue(dictionary === obj.mapAny)
+    }
+
+    #if false
+    func testValueForKeyLinkingObjects() {
```
Comment (leemaguire) [inline RealmSwift/Tests/ModernObjectTests.swift:258]: Is this planned to be used?

File: RealmSwift/Util.swift [hunk 2]
```diff
@@ -75,18 +75,37 @@ extension ObjectBase {
...
 // MARK: CustomObjectiveCBridgeable
 
 /// :nodoc:
 public func dynamicBridgeCast<T>(fromObjectiveC x: Any) -> T {
-    if T.self == DynamicObject.self {
+    return failableDynamicBridgeCast(fromObjectiveC: x)!
+}
+
+/// :nodoc:
+public func failableDynamicBridgeCast<T>(fromObjectiveC x: Any) -> T? {
```
Comment (leemaguire) [inline RealmSwift/Util.swift:98]: should this and `dynamicBridgeCast` not be prefixed with `_`

File: Realm/RLMSet.mm [hunk 3]
```diff
@@ -259,6 +259,18 @@ - (void)removeObject:(id)object {
...
 }
 
+- (void)replaceAllObjectsWithObjects:(NSArray *)objects {
+    changeSet(self, ^{
+        [_backingCollection removeAllObjects];
+        if (!objects || (id)objects == NSNull.null) {
+            return;
+        }
+        for (id object in objects) {
+            [_backingCollection addObject:object];
```
[the remark this answers is not in the document]
Comment (tgoyne) [reply Realm/RLMSet.mm:269]: `objects` isn't actually always a NSArray and `addObjectsFromArray:` specifically requires an NSArray and not any enumerable.

File: Realm/RLMObjectBase.mm [hunk 4]
```diff
@@ -747,7 +755,13 @@ uint64_t RLMObjectBaseGetCombineId(__unsafe_unretained RLMObjectBase *const obj)
 }
 
 @implementation RealmSwiftObject
++ (BOOL)accessInstanceVariablesDirectly {
```
[the remark this answers is not in the document]
Comment (tgoyne) [reply Realm/RLMObjectBase.mm:758]: This is called by KVO to decide how to get the old/new values. Without this it tries to read the ivars for the PW and explodes.

File: Realm/RLMArray.mm [hunk 5]
```diff
@@ -38,13 +38,11 @@ @implementation RLMArrayHolder
 @end
 
 @interface RLMArray () <RLMThreadConfined_Private>
+// Backing array when this instance is unmanaged
+@property NSMutableArray *backingCollection;
```
[the remark this answers is not in the document]
Comment (tgoyne) [reply Realm/RLMArray.mm:42]: There was something which used this property to work generically on both sets and arrays, but maybe I ended up removing that and if I did this should go too.

File: RealmSwift/Impl/BasicTypes.swift [hunk 6]
```diff
@@ -19,74 +19,302 @@
...
+
+// MARK: - Modern property getters/setters
+
+public protocol _Int: BinaryInteger, _OptionalPersistable, _DefaultConstructible, _PrimaryKey, _Indexable {
+}
+
+extension _Int {
+    @inlinable
+    public static func _rlmGetProperty(_ obj: ObjectBase, _ key: PropertyKey) -> Self {
+        return Self(RLMGetSwiftPropertyInt64(obj, key))
```
[the remark this answers is not in the document]
Comment (tgoyne) [reply RealmSwift/Impl/BasicTypes.swift:117]: I initially had them but it doesn't actually simplify anything or have any functional benefits.

File: RealmSwift/Util.swift (hunk 2 above)
Comment (tgoyne) [reply RealmSwift/Util.swift:98]: I'm not sure why dynamicBridgeCast is public at all. I'll make the new function internal, though.

File: Realm/RLMAccessor.mm [hunk 7]
```diff
@@ -30,18 +30,17 @@
...
+#import "RLMUtil.hpp"
 
 #import <realm/object-store/results.hpp>
 #import <realm/object-store/property.hpp>
 
 #import <objc/runtime.h>
 #import <objc/message.h>
 
-#pragma mark - Helper functions
+#pragma mark Helper functions
```
[the remark this answers is not in the document]
Comment (tgoyne) [reply Realm/RLMAccessor.mm:43]: It's there to add a separator line in the dropdown, and this is at the very top of the dropdown.

File: RealmSwift/Impl/SchemaDiscovery.swift [hunk 8]
```diff
@@ -19,52 +19,76 @@
...
-private func getProperties(_ cls: RLMObjectBase.Type) -> [RLMProperty] {
-    let object = cls.init()
+private func getModernProperties(_ object: ObjectBase) -> [RLMProperty] {
+    return Mirror(reflecting: object).children.compactMap { prop in
+        guard let label = prop.label else { return nil }
+        guard let value = prop.value as? DiscoverablePersistedProperty else {
+            return nil
+        }
+        let property = RLMProperty(name: label, value: value)
+        property.swiftIvar = ivar_getOffset(class_getInstanceVariable(type(of: object), label)!)
```
Comment (jsflax) [inline RealmSwift/Impl/SchemaDiscovery.swift:71]: Once we have the obj key and col key, why do we still need the ivar offset?

File: Realm/RLMArray.mm [hunk 9]
```diff
@@ -517,9 +533,12 @@ - (NSArray *)objectsAtIndexes:(NSIndexSet *)indexes {
 
 - (BOOL)isEqual:(id)object {
     if (auto array = RLMDynamicCast<RLMArray>(object)) {
-        return !array.realm
-        && ((_backingCollection.count == 0 && array->_backingCollection.count == 0)
-            || [_backingCollection isEqual:array->_backingCollection]);
+        if (array.realm) {
+            return NO;
```
[the remark this answers is not in the document]
Comment (tgoyne) [reply Realm/RLMArray.mm:537]: This array is unmanaged, and unmanaged arrays are never equal to managed ones.

File: Realm/RLMClassInfo.hpp [hunk 10]
```diff
@@ -96,13 +96,16 @@ class RLMClassInfo {
     RLMClassInfo &resolve(RLMRealm *);
 
     // Return true if the RLMObjectSchema is for a Swift class
-    bool isSwiftClass();
+    bool isSwiftClass() const noexcept;
+
+    // Returns true if this was a dynamically added type
+    bool isDynamic() const noexcept;
```
[the remark this answers is not in the document]
Comment (tgoyne) [reply Realm/RLMClassInfo.hpp:102]: Types which are not in the schema which were added when resolving typed links and use the dynamicObjectSchema members of this class.

File: RealmSwift/Impl/Persistable.swift [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+// the property's index in the object schema, but that's not something that any
+// of the Swift code should rely on. In the future it may make sense to change
+// this to the ColKey.
+public typealias PropertyKey = UInt16
+
+// A tag protocol used in schema discovery to find @Persisted properties
+internal protocol DiscoverablePersistedProperty: _RealmSchemaDiscoverable {}
+
+// A type which can be stored by the @Persisted property wrapper
+public protocol _Persistable: _RealmSchemaDiscoverable {
```
[the remark this answers is not in the document]
Comment (tgoyne) [reply RealmSwift/Impl/Persistable.swift:32]: No, the API for that needs to be fairly different and is probably just going to be mappings between types to be able to support collections.

File: RealmSwift/Impl/SchemaDiscovery.swift (hunk 8 above)
Comment (tgoyne) [reply RealmSwift/Impl/SchemaDiscovery.swift:71]: To read from and write to unmanaged properties from obj-c.

Comment (tgoyne) [reply RealmSwift/Impl/SchemaDiscovery.swift:71]: Also very importantly this is how we set the col key on the managed property wrappers.

**Review by jsflax — APPROVED** [review]
Awesome 🤩. This is a very exciting time for the codebase. Updates on the README are in our near future. The pull request LGTM. @leemaguire would you mind doing one more pass on this as well?

**Review by leemaguire — APPROVED** [review]
🎉

**Pull request merged**